File: 2699-count-the-number-of-fair-pairs/2699-count-the-number-of-fair-pairs.java
class Solution {
    public long countFairPairs(int[] nums, int lower, int upper) {
        Arrays.sort(nums);
        long ans = 0;
        for(int i=0;i+1<nums.length;i++){
            int nl = lower - nums[i];
            int nu = upper - nums[i];
            int low = lowerBound(nums, i + 1, nums.length, nl);
            int high = upperBound(nums, i + 1, nums.length, nu);
            ans += (high - low);
        }
        return ans;
    }

    private int lowerBound(int[] nums, int start, int end, int target) {
        int lo = start, hi = end;
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (nums[mid] < target) lo = mid + 1;
            else hi = mid;
        }
        return lo;
    }

    private int upperBound(int[] nums, int start, int end, int target) {
        int lo = start, hi = end;
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (nums[mid] <= target) lo = mid + 1;
            else hi = mid;
        }
        return lo;
    }
}

File: 2699-count-the-number-of-fair-pairs/2699-count-the-number-of-fair-pairs.py
class Solution:
    def countFairPairs(self, nums: List[int], lower: int, upper: int) -> int:
        nums.sort()
        count = 0
        n = len(nums)

        for i in range(n):
            low_val = lower - nums[i]
            high_val = upper - nums[i]

            l = i + 1
            while l < n and nums[l] < low_val:
                l += 1

            r = l
            while r < n and nums[r] <= high_val:
                r += 1

            count += (r - l)

        return count
